core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -301,9 +337,37 @@ private boolean assertSerialization(DocumentMapper mapper) {
         return true;
     }
 
-    protected void checkNewMappersCompatibility(Collection<ObjectMapper> newObjectMappers, Collection<FieldMapper> newFieldMappers, boolean updateAllTypes) {
+    /**
+     * Check that fields are defined only once. It is possible to define fields
+     * several times eg. if a field is defined once via 'fields' in the mapping
+     * and once from a field mapper.
+     */
+    private void checkUniqueness(Collection<ObjectMapper> objectMappers, Collection<FieldMapper> fieldMappers) {
```

Comment:
Or is this the object/field check + a nicer error message if you define twice (which would otherwise fail when trying to update the mappings?)? Or was the latter not being enforced?